File: mobile/library/kotlin/src/io/envoyproxy/envoymobile/GRPCRequestBuilder.kt
package io.envoyproxy.envoymobile

/**
 * Builder used for creating new gRPC `Request` instances.
 */
class GRPCRequestBuilder(
    path: String,
    authority: String,
    useHTTPS: Boolean
) {
  private val underlyingBuilder: RequestBuilder = RequestBuilder(
      method = RequestMethod.POST,
      scheme = if (useHTTPS) "https" else "http",
      authority = authority,
      path = path)

  init {
    underlyingBuilder.addHeader("content-type", "application/grpc")
  }

  /**
   * Append a value to the header key.
   *
   * @param name the header key.
   * @param value the value associated to the header key.
   * @return this builder.
   */
  fun addHeader(name: String, value: String): GRPCRequestBuilder {
    underlyingBuilder.addHeader(name, value)
    return this
  }

  /**
   * Remove the value in the specified header.
   *
   * @param name the header key to remove.
   * @param value the value to be removed.
   * @return this builder.
   */
  fun removeHeader(name: String, value: String): GRPCRequestBuilder {
    underlyingBuilder.removeHeader(name, value)
    return this
  }

  /**
   * Remove all headers with this name.
   *
   * @param name the header key to remove.
   * @return this builder.
   */
  fun removeHeaders(name: String): GRPCRequestBuilder {
    underlyingBuilder.removeHeaders(name)
    return this
  }

  /**
   * Add a specific timeout for the gRPC request. This will be sent in the `grpc-timeout` header.
   *
   * @param timeoutMS Timeout, in milliseconds.
   * @return this builder.
   */
  fun addTimeoutMS(timeoutMS: Int?): GRPCRequestBuilder {
    val headerName = "grpc-timeout"
    if (timeoutMS == null) {
      removeHeaders(headerName)
    } else {
      addHeader(headerName, "${timeoutMS}m")
    }
    return this
  }

  /**
   * Creates the Request object using the data set in the builder.
   *
   * @return the Request object.
   */
  fun build(): Request {
    return underlyingBuilder.build()
  }
}

File: mobile/library/swift/src/grpc/GRPCRequestBuilder.swift
import Foundation

/// Builder used for creating new gRPC `Request` instances.
@objcMembers
public final class GRPCRequestBuilder: NSObject {
  private let underlyingBuilder: RequestBuilder

  /// Initialize a new builder.
  ///
  /// - parameter path:      Path of the RPC (i.e., `/pb.api.v1.Foo/GetBar`).
  /// - parameter authority: Authority to use for the RPC (i.e., `api.foo.com`).
  /// - parameter useHTTPS:  Whether to use HTTPS (or HTTP).
  public init(path: String, authority: String, useHTTPS: Bool = true) {
    self.underlyingBuilder = RequestBuilder(method: .post,
                                            scheme: useHTTPS ? "https" : "http",
                                            authority: authority,
                                            path: path)
    self.underlyingBuilder.addHeader(name: "content-type", value: "application/grpc")
    self.underlyingBuilder.addUpstreamHttpProtocol(.http2)
  }

  /// Append a value to the header key.
  ///
  /// - parameter name:  The header key.
  /// - parameter value: Value the value associated to the header key.
  ///
  /// - returns: This builder.
  @discardableResult
  public func addHeader(name: String, value: String) -> GRPCRequestBuilder {
    self.underlyingBuilder.addHeader(name: name, value: value)
    return self
  }

  /// Remove all headers with this name.
  ///
  /// - parameter name: The header key to remove.
  ///
  /// - returns: This builder.
  @discardableResult
  public func removeHeaders(name: String) -> GRPCRequestBuilder {
    self.underlyingBuilder.removeHeaders(name: name)
    return self
  }

  /// Remove the value in the specified header.
  ///
  /// - parameter name: The header key to remove.
  /// - parameter value: The value to be removed.
  ///
  /// - returns: This builder.
  @discardableResult
  public func removeHeader(name: String, value: String) -> GRPCRequestBuilder {
    self.underlyingBuilder.removeHeader(name: name, value: value)
    return self
  }

  /// Add a specific timeout for the gRPC request. This will be sent in the `grpc-timeout` header.
  ///
  /// - parameter timeoutMS: Timeout, in milliseconds.
  ///
  /// - returns: This builder.
  @discardableResult
  public func addTimeoutMS(_ timeoutMS: UInt?) -> GRPCRequestBuilder {
    let headerName = "grpc-timeout"
    if let timeoutMS = timeoutMS {
      self.addHeader(name: headerName, value: "\(timeoutMS)m")
    } else {
      self.removeHeaders(name: headerName)
    }

    return self
  }

  /// Creates a request object from the builder.
  ///
  /// - returns: The new request object.
  public func build() -> Request {
    return self.underlyingBuilder.build()
  }
}
